### PR TITLE
Adjust buildRecordHashFunction to skip param fields and type fields

### LINKED ISSUE #11613: 'chpl__defaultHash' on generic records attempt to hash 'type' fields and fail
If you have `R` as generic, the `chpl__defaultHash` will try to hash *all* fields including a `type` field.

```chpl
record R {
	type t;
	var a : [1..10] t;
}
proc ==(a: R, b: R) {
  return && reduce (a.a == b.a);
}
proc !=(a: R, b: R) {
  return || reduce (a.a != b.a);
}

var d : domain(R(int));
d += new R(int);

writeln(d);
```
[TIO](https://tio.run/##dc5BCoMwEAXQtTnFdyMJLaFuFS/htnQxJoEKNsoQlVJ79jRSaOmim2H4Mw@@udLkhhjZmZEtWjxEFu6TQ6hFthCDUOFcal2eLnv2FBOPBk0jqUJ7RJemSghgF2b2KIq02dk4SNKUHtFpUh@Y/4Xb9gPzLxR7D5t62PFGvZet7H1Q6WRxaODdindSC7FyH9zgpVV1jC8)

**Output**

```
error: unresolved call 'chpl__defaultHash(type int(64))'
.code.tio.chpl:1: note: candidates are: chpl__defaultHash(r: R)
$CHPL_HOME/modules/internal/String.chpl:2097: note:                 chpl__defaultHash(x: string)
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:811: note:                 chpl__defaultHash(b: bool)
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:818: note:                 chpl__defaultHash(i: int(64))
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:822: note:                 chpl__defaultHash(u: uint(64))
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:826: note:                 chpl__defaultHash(e)
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:830: note:                 chpl__defaultHash(f: real)
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:834: note:                 chpl__defaultHash(c: complex)
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:838: note:                 chpl__defaultHash(a: imag)
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:842: note:                 chpl__defaultHash(u: chpl_taskID_t)
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:846: note:                 chpl__defaultHash(l: [] )
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:856: note:                 chpl__defaultHash(o: borrowed object)
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:864: note:                 chpl__defaultHash(r: range)
$CHPL_HOME/modules/internal/String.chpl:182: note:                 chpl__defaultHash(r: chpl__inPlaceBuffer)
$CHPL_HOME/modules/internal/String.chpl:206: note:                 chpl__defaultHash(r: __serializeHelper)
$CHPL_HOME/modules/internal/String.chpl:231: note:                 chpl__defaultHash(r: codePointIndex)
$CHPL_HOME/modules/internal/OwnedObject.chpl:186: note:                 chpl__defaultHash(r: _owned)
$CHPL_HOME/modules/internal/SharedObject.chpl:138: note:                 chpl__defaultHash(r: _shared)
$CHPL_HOME/modules/internal/Atomics.chpl:176: note:                 chpl__defaultHash(r: AtomicBool)
$CHPL_HOME/modules/internal/Atomics.chpl:314: note:                 chpl__defaultHash(r: AtomicT)
$CHPL_HOME/modules/internal/NetworkAtomics.chpl:32: note:                 chpl__defaultHash(r: RAtomicBool)
$CHPL_HOME/modules/internal/NetworkAtomics.chpl:117: note:                 chpl__defaultHash(r: RAtomicT)
$CHPL_HOME/modules/internal/AtomicsCommon.chpl:23: note:                 chpl__defaultHash(r: atomic_refcnt)
$CHPL_HOME/modules/internal/ChapelTuple.chpl:43: note:                 chpl__defaultHash(r: _tuple)
$CHPL_HOME/modules/internal/ChapelSyncvar.chpl:108: note:                 chpl__defaultHash(r: _syncvar)
$CHPL_HOME/modules/internal/ChapelSyncvar.chpl:633: note:                 chpl__defaultHash(r: _singlevar)
$CHPL_HOME/modules/internal/ChapelSyncvar.chpl:886: note:                 chpl__defaultHash(r: chpl_sync_aux_t)
$CHPL_HOME/modules/internal/ChapelSyncvar.chpl:912: note:                 chpl__defaultHash(r: chpl_single_aux_t)
$CHPL_HOME/modules/internal/ChapelLocale.chpl:485: note:                 chpl__defaultHash(r: localesBarrier)
$CHPL_HOME/modules/internal/DefaultRectangular.chpl:700: note:                 chpl__defaultHash(r: _remoteAccessData)
$CHPL_HOME/modules/internal/ChapelArray.chpl:898: note:                 chpl__defaultHash(r: dmap)
$CHPL_HOME/modules/internal/ChapelArray.chpl:924: note:                 chpl__defaultHash(r: _distribution)
$CHPL_HOME/modules/internal/ChapelArray.chpl:1040: note:                 chpl__defaultHash(r: _serialized_domain)
$CHPL_HOME/modules/internal/ChapelArray.chpl:1057: note:                 chpl__defaultHash(r: _domain)
$CHPL_HOME/modules/internal/ChapelArray.chpl:3986: note:                 chpl__defaultHash(r: _OpaqueIndex)
$CHPL_HOME/modules/internal/LocaleTree.chpl:29: note:                 chpl__defaultHash(r: chpl_localeTreeRecord)
$CHPL_HOME/modules/internal/DefaultAssociative.chpl:36: note:                 chpl__defaultHash(r: chpl_TableEntry)
$CHPL_HOME/modules/internal/ExternalArray.chpl:28: note:                 chpl__defaultHash(r: chpl_external_array)
$CHPL_HOME/modules/internal/ChapelError.chpl:106: note:                 chpl__defaultHash(r: chpl_TaskErrors)
$CHPL_HOME/modules/internal/LocaleModelHelpSetup.chpl:44: note:                 chpl__defaultHash(r: chpl_root_locale_accum)
$CHPL_HOME/modules/internal/LocaleModelHelpRuntime.chpl:38: note:                 chpl__defaultHash(r: chpl_localeID_t)
$CHPL_HOME/modules/internal/LocaleModelHelpRuntime.chpl:45: note:                 chpl__defaultHash(r: chpl_comm_on_bundle_t)
$CHPL_HOME/modules/internal/LocaleModelHelpRuntime.chpl:48: note:                 chpl__defaultHash(r: chpl_task_bundle_t)
$CHPL_HOME/modules/standard/CommDiagnostics.chpl:176: note:                 chpl__defaultHash(r: chpl_commDiagnostics)
$CHPL_HOME/modules/standard/List.chpl:53: note:                 chpl__defaultHash(r: list)
$CHPL_HOME/modules/standard/IO.chpl:750: note:                 chpl__defaultHash(r: iostyle)
$CHPL_HOME/modules/standard/IO.chpl:1061: note:                 chpl__defaultHash(r: qio_conv_t)
$CHPL_HOME/modules/standard/IO.chpl:1256: note:                 chpl__defaultHash(r: file)
$CHPL_HOME/modules/standard/IO.chpl:1905: note:                 chpl__defaultHash(r: channel)
$CHPL_HOME/modules/standard/IO.chpl:2026: note:                 chpl__defaultHash(r: ioChar)
$CHPL_HOME/modules/standard/IO.chpl:2058: note:                 chpl__defaultHash(r: ioNewline)
$CHPL_HOME/modules/standard/IO.chpl:2088: note:                 chpl__defaultHash(r: ioLiteral)
$CHPL_HOME/modules/standard/IO.chpl:2112: note:                 chpl__defaultHash(r: ioBits)
$CHPL_HOME/modules/standard/IO.chpl:4398: note:                 chpl__defaultHash(r: ItemReader)
$CHPL_HOME/modules/standard/IO.chpl:4440: note:                 chpl__defaultHash(r: ItemWriter)
$CHPL_HOME/modules/standard/Sys.chpl:215: note:                 chpl__defaultHash(r: sys_sockaddr_t)
$CHPL_HOME/modules/standard/Sys.chpl:224: note:                 chpl__defaultHash(r: sys_addrinfo_t)
$CHPL_HOME/modules/standard/Regexp.chpl:344: note:                 chpl__defaultHash(r: qio_regexp_options_t)
$CHPL_HOME/modules/standard/Regexp.chpl:384: note:                 chpl__defaultHash(r: qio_regexp_string_piece_t)
$CHPL_HOME/modules/standard/Regexp.chpl:506: note:                 chpl__defaultHash(r: stringPart)
$CHPL_HOME/modules/standard/Regexp.chpl:527: note:                 chpl__defaultHash(r: reMatch)
$CHPL_HOME/modules/standard/Regexp.chpl:573: note:                 chpl__defaultHash(r: regexp)
```

**Note: Long error messages like these are kind of why I want improved 'candidate' messages as mentioned in #11611**

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Adjust buildRecordHashFunction to skip param fields and type fields (#17533)

Resolves #11613.

Adjust the compiler code which generates `chpl__defaultHash` for
record types so that it does not hash type fields or param fields.

Skip type fields because they do not meaningfully differentiate the
final hash that is produced. Similarly, param fields will be folded
away and so do not need to be hashed.

Reviewed by bradcray. Thanks!

TESTING:

- [x] `ALL` on `linux64` when `COMM=none`
- [x] `ALL` on `linux64` when `COMM=gasnet`

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

File: compiler/passes/buildDefaultFunctions.cpp [hunk 1]
```diff
@@ -1592,8 +1596,15 @@ static void buildRecordHashFunction(AggregateType *ct) {
       }
       i++;
     }
-    fn->insertAtTail(new CallExpr(PRIM_RETURN, call));
+
+    if (call) {
+      fn->insertAtTail(new CallExpr(PRIM_RETURN, call));
```
Comment (bradcray) [inline compiler/passes/buildDefaultFunctions.cpp:1601]: What's the situation in which we'd hit `call == NULL`?  When all fields are types and params?

Comment (bradcray) [reply compiler/passes/buildDefaultFunctions.cpp:1601]: (and, assuming so, does your test case cover that code path?  I didn't see a case that looked like it did, but am reading fast).

Comment (dlongnecke-cray) [reply compiler/passes/buildDefaultFunctions.cpp:1601]: Yeah. One example of that happening is the base `_tuple` type before we specialize it.

Comment (dlongnecke-cray) [reply compiler/passes/buildDefaultFunctions.cpp:1601]: I'll add a test for that case, thanks!

Comment (dlongnecke-cray) [reply compiler/passes/buildDefaultFunctions.cpp:1601]: Added in d359a82309f8a404a3e911f2061a6cce376ac153.

**Review by bradcray — APPROVED** [review]

**Comment by bradcray** [comment]
Looks good to me, thanks!

**Pull request merged**